r/R/trackplots.R

```diff
@@ -113,6 +113,92 @@ get_trackplot_height <- function(plot) {
   }
 }
 
+#' Calculate y positions for transcript segments to avoid overlap
+#' Steps:
+#' 1. Calculate the maximum overlap depth of transcripts
+#' 2. Iterate through transcript start/end in sorted order
+#' 3. Randomly assign each transcript a y-coordinate between 1 and max overlap depth,
+#'   with the restriction that a transcript can't have the same y-coordinate
+#' If is a transcript, filter out the exons in determining overlap so there aren't repeated steps
+trackplot_calculate_segment_height <- function(data, is_transcript = FALSE) {
+  data <- tibble::rownames_to_column(data, "idx_col")
+  boundaries <- data %>%
+    dplyr::filter(if(is_transcript) {!!as.symbol("feature") == "transcript"} else {TRUE}) %>%
+    {
+      dplyr::bind_rows(
+        dplyr::transmute(., pos = start, idx_col, start = TRUE),
+        dplyr::transmute(., pos = end, idx_col, start = FALSE)
+      )
+    } %>%
+    dplyr::arrange(pos)
+  
+  if (nrow(data) > 0) total_positions <- max(cumsum(2 * boundaries$start - 1))
+  else total_positions <- 0L
+  y_pos <- integer(0) # Names: tx_id, value: y position for transcript
+  occupied_y <- rep(FALSE, total_positions) # Boolean vector, marking which y positions are open
+  prev_seed <- get_seed()
+  set.seed(12057235)
+  free_positions <- total_positions
+  for (i in seq_len(nrow(boundaries))) {
+    row <- boundaries[i, ]
+    if (row$start) {
+      # if else block to account for sample interpreting int of length 1 as a range
+      if (free_positions == 1) {
+        assigned_pos <- which(!occupied_y)[1]
+      } else {
+        assigned_pos <- sample(which(!occupied_y), 1)
+      }
```

Comment:
I thought about that one too but I was worried of the case of having a large region with many peaks.  Since length is an O(n) check rather than an o(1) check it might end up taking up unnecessary compute.  Stylistically this solution is nicer though.